### ISSUE TITLE
smart-garden-gateway-mqtt-client and openHAP/HA integrations

### ISSUE THREAD

**Issue by easybe** [andrexp/gardena-local-control#25]
As you might know, I (a GARDENA smart Gateway maintainer) have been working on something similar to this project. It comes in the form of an add-on package that can be installed over serial/shell access. The installation instructions can be found [here](https://github.com/easybe/smart-garden-gateway-mqtt-client).

As it only covers the MQTT to NNG part and shall not implement an API, some more work would be needed to create a more user-friendly solution. Would you perhaps be willing to collaborate on some sort of integration for Home Assistant or openHAP? 

As we have learned in [this conversation](https://github.com/openhab/openhab-addons/issues/14904) the openHAP people don't really seem to be interested in adapting their integration to use MQTT. Should we ask the [Home Assistant people](https://github.com/py-smart-gardena/hass-gardena-smart-system) what solution they can imagine? I guess the HA integration should be rather easy to extend as it is Python.

Further, I was wondering if some sort of library for the local API would be possible that then can be used by the different home automation solutions. Any thoughts are welcome.

**Comment by easybe** [comment]
@andrexp, as [your comment](https://github.com/openhab/openhab-addons/issues/14904#issuecomment-2453546636) is no longer related to openHAB, I would rather move the conversation here. As I use Home Assistant as well, I would be happy to support you with your efforts. However, we need to put some thought into the implementation and carefully consider the different options. 

At this point, I do not know what the best solution is and don't have the answers you are seeking, sorry.

**Comment by andrexp** [comment]
Ouch... I just opened a new issue without recognizing yours - please ignore it. Lets continue here.

As I previously discussed [here](https://github.com/openhab/openhab-addons/issues/14904#issuecomment-2453653561) I need a few more information about the (possible) NNG-Stream commands. Is there any command to discover all devices which are connected to the Gardena Smart Gateway? And: Is there a kind of type encoded into the Information which gives us the information about the device type (mower, irrigation control........)

I could imagine to add the support for the MQTT auto discovery feature first to support at least the basic functions as they are would already work if one configures the mqtt entities manually. This has the advantage being supported by openHAB as well, as the HA-discovery-messages are also being recognized by the MQTT binding.

At this time I am not sure if I could spend the time to create a whole (semi-)official integration although I would have a lot of fun learning and doing so. I will do some research tomorrow and see how much work this would be. Meanwhile you may can answer my questions ;-) Thank you in advance.

**Comment by easybe** [comment]
> Is there any command to discover all devices which are connected to the Gardena Smart Gateway? 

I don't think so, I will need to do some more investigation in order to answer that question. Anyway, it should be possible to implement that feature in  [smart-garden-gateway-mqtt-client ](https://github.com/easybe/smart-garden-gateway-mqtt-client).

> And: Is there a kind of type encoded into the Information which gives us the information about the device type (mower, irrigation control........)

The devices provide that information at startup, e.g.:
```
    "model_number": {
        "ts": 1731064733,
        "vs": "19040"
    },
    "device_type": {
        "ts": 1731064733,
        "vs": "Sensor"
    },
```

> I could imagine to add the support for the MQTT auto discovery feature first to support at least the basic functions as they are would already work if one configures the mqtt entities manually. This has the advantage being supported by openHAB as well, as the HA-discovery-messages are also being recognized by the MQTT binding.

I think that is something we could implement in [smart-garden-gateway-mqtt-client ](https://github.com/easybe/smart-garden-gateway-mqtt-client) too.

**Comment by ETE-Design** [comment]
@easybe & @andrexp Any progress? Would love to help you test it out - Though I only have one Lawn Mover connected to my Gardena Smart Gateway...

**Comment by andrexp** [comment]
Sorry for the late response. As you can imagine, the planned research has still not taken place. However, it is still on my list. Until then, I have added the MQTT auto-discovery functionality to my local code for some testing which is working so far. I will share it before the upcoming lawnmower season.

**Comment by ETE-Design** [comment]
> Sorry for the late response. As you can imagine, the planned research has still not taken place. However, it is still on my list. Until then, I have added the MQTT auto-discovery functionality to my local code for some testing which is working so far. I will share it before the upcoming lawnmower season.

I'm not in a hurry - Would just like to integrate in my Home Assistant without the need of their cloud... If you need me to help you test anything then let me know :-) I'm not a programmer but gladly help you test anything needed.

**Comment by andrexp** [comment]
I just want to let you know I am currently try to develop a homeassistant integration. Unfortunately it is a bit more complicated as I thought, so it will need some time to finish. As soon as I have a prototype I will let you know

**Comment by ETE-Design** [comment]
@andrexp That is amazing news :-)

**Comment by mija87** [comment]
Exciting! Follow!

**Comment by ETE-Design** [comment]
@andrexp Any progress? :-D

**Comment by andrexp** [comment]
Unfortunately, no. As soon as I have anything runnable I will let you know. Anyway nice to hear that the interest is still out there... ;-)

**Comment by easybe** [comment]
Some of us at GARDENA are planning on putting some effort into improving the experience for the home automation community. How exactly that will look like is still open though. I will keep you updated and reach out to make sure we are going into the right direction.

**Comment by ETE-Design** [comment]
@easybe Would be nice with a way to control it al local ;-) Any progress on that?

@andrexp Did you manage to get something up and running?

**Comment by andrexp** [comment]
Hi, my companion @FredBit90 had some messages with the Gardena devs. Afaik they also work on something so I stopped my stuff for now. But the main reason was the lack of time. Do hopefully the "professional" solution will not take too long. Sorry that I have no good news.

**Comment by easybe** [comment]
Indeed, we at GARDENA are working on a Python library and a HA integration for local control here: https://github.com/cloudless-garden

However, everything is still at a very early stage and not really usable yet.

If you let us know which (exact) products you own we might be able to add support for them sooner than later.

**Comment by ETE-Design** [comment]
@easybe I own a Smart Sileno+, will be more than happy to help you test it against my device if needed?

**Comment by easybe** [comment]
> [@easybe](https://github.com/easybe) I own a Smart Sileno+, will be more than happy to help you test it against my device if needed?

OK. Could you describe your use case a bit? How does your current setup look like? Which features do you need?

**Comment by ETE-Design** [comment]
@easybe - Running Home Assistant OS on a Raspberry Pi 5 - 8 GB

Some features I can think of could be:

Firmware update
Manual control
Manual Start / Stop
Return to base
Return to base if battery is under xx%
Schedule cut
Return to base if battery is under xx% - Then continue schedule when charged
Cut for x Min/Hour
Animal friendly Cut: sunrise to sunset
Stop cutting if rain
Depending on what function the Smart Sileno+ expose to you now that you have more background knowhow a "Spot Cut" so it start in the area you manual control it to and start cutting in a radius like 2 meters
A way to make it cut smarter like in stripes so when it sense the wire it turn 90 degree move x cm turn again 90 degree and move next the line already cut. 
I'm sure I can find more features I would like to have over time.

I know some of these functions can be done with automations in HA depending on what entitles you make for us. But to do it as user friendly as possible mabye a custom lovelace card would be the best way to present it for users.

**Comment by easybe** [comment]
OK, thanks for the info.

Do you have anything of this already set up in HA with the currently available integration that goes through the cloud?

At least for now, we will only be able to provide basic control/monitoring. Any logic needs to be implemented through automations. But, we'll have to see what makes sense in the future.

**Comment by andrexp** [comment]
@FredBit90 an me own a Slieno Life (correct me, if I am wrong @FredBit90).

We are also very interested in basic functionality. With our script it was basically start/stop the mower and manual mowing for x minutes. The schedule was managed through automations. I also added some useful features like stop mowing when my weather sensor has detected rain and so on... 

Long story short: it would be great to have a local connection with HA as soon as possible :-)

**Comment by easybe** [comment]
> [@FredBit90](https://github.com/FredBit90) an me own a Slieno Life (correct me, if I am wrong [@FredBit90](https://github.com/FredBit90)).
> 
> We are also very interested in basic functionality. With our script it was basically start/stop the mower and manual mowing for x minutes. The schedule was managed through automations. I also added some useful features like stop mowing when my weather sensor has detected rain and so on...
> 
> Long story short: it would be great to have a local connection with HA as soon as possible :-)

That should be achievable in the next few weeks (with our WIP HACS integration: https://github.com/cloudless-garden/ha-gardena-smart-local-preview).

I guess we should implement this: https://www.home-assistant.io/integrations/lawn_mower/

**Comment by andrexp** [comment]
Yes, this will be fine.

In my approach to create an HA Integration, I used the MQTT client you have written for the Gardena gateway. I edited it in a way that, if new devices are discovered on lemonbeat, the client sends an retained message with basic information of the lawn mower (or more generic: the discovered Gardena device). We should implement this as well. HA integration often work exactly like this. I think in the dev papers it was called: config flow. Maybe you as dev know a command which can be used that the Gardena gateway exposes all of the connected devices or we implement a feature like my approach above. What do you think?

**Comment by easybe** [comment]
> Yes, this will be fine.
> 
> In my approach to create an HA Integration, I used the MQTT client you have written for the Gardena gateway. I edited it in a way that, if new devices are discovered on lemonbeat, the client sends an retained message with basic information of the lawn mower (or more generic: the discovered Gardena device). We should implement this as well. HA integration often work exactly like this. I think in the dev papers it was called: config flow. Maybe you as dev know a command which can be used that the Gardena gateway exposes all of the connected devices or we implement a feature like my approach above. What do you think?

We have that part already sorted. We query the gateway for devices on startup.

We are also working on automatic gateway discovery through Zeroconf: https://github.com/cloudless-garden/ha-gardena-smart-local-preview/pull/6

Regarding mowers, all we have so far is some low-level support: https://github.com/cloudless-garden/gardena-smart-local-api/pull/11

**Comment by andrexp** [comment]
Sounds great!!

I should download your code and try myself a bit around. As far as I understand those snippets, the most important device description with all commands previously sent on nng communication is implemented. Yes, it should be put all together into the HA lawn mower object, but anyway: great work and I am happy to see that - thank you so much for sharing this!

**Comment by easybe** [comment]
> Sounds great!!
> 
> I should download your code and try myself a bit around. As far as I understand those snippets, the most important device description with all commands previously sent on nng communication is implemented. Yes, it should be put all together into the HA lawn mower object, but anyway: great work and I am happy to see that - thank you so much for sharing this!

Currently, you still need an unreleased image for the gateway. But as you have shell access, you could easily install it.

On a second generation gateway (art. no. 19005):
```
fw_setenv update_url http://gateway-dev.iot.sg.dss.husqvarnagroup.net/feeds/qa/images/gardena-update-image-bnw-zephyr-gardena-sg-mt7688.swu
systemctl restart swupdate-check
```

**Comment by ETE-Design** [comment]
> OK, thanks for the info.
> 
> Do you have anything of this already set up in HA with the currently available integration that goes through the cloud?
> 
> At least for now, we will only be able to provide basic control/monitoring. Any logic needs to be implemented through automations. But, we'll have to see what makes sense in the future.

I haven't tryed using the cloud cause anything else on my Home Assistant run local - So for now I have only used the APP from my phone.

Any chance you will release a "Beta Firmware" for the gateway so we can start trying out what your doing ASAP?

**Comment by easybe** [comment]
> Any chance you will release a "Beta Firmware" for the gateway so we can start trying out what your doing ASAP?

I shared the needed commands to install a "beta" image one message up, but, you currently still need serial access for that, sorry.

Hopefully the official release will be rolled out in 3.5 weeks. By then the integration stuff should have advanced too. Anyway, at this very point the available integration is not really usable yet.

**Comment by ETE-Design** [comment]
@easybe I see that you have added Lawn Mower's - Is there a guide on how to enable shell access from Windows? Or do you know the status on the Official Firmware for this?

**Comment by andrexp** [comment]
You need to open the gateway and a USB-UART- cable or something similar.

The official manual to gain access to the gateway's serial port is found in our description/FAQ or directly here: https://github.com/husqvarnagroup/smart-garden-gateway-public?tab=readme-ov-file#uart

After you have serial access you will be able to add some SSH keys and manage further connections over TCP/SSH.

**Comment by easybe** [comment]
> [@easybe](https://github.com/easybe) I see that you have added Lawn Mower's - Is there a guide on how to enable shell access from Windows? Or do you know the status on the Official Firmware for this?

Best just be patient for another 1.5 weeks. And if you then wait another 3 weeks, it will be easier, you won’t even need to use SSH.

**Comment by easybe** [comment]
Gateway software version 10.4.3 has been rolled out 🍾️

Enabling SSH through the HTTP API should now work as described [here](https://github.com/husqvarnagroup/smart-garden-gateway-public#ssh).

On Windows, you can try the following in PowerShell.

If you don't already have one, create an SSH key pair:
```txt
ssh-keygen
```

If you do not have SSH installed, try:
```powershell
Add-WindowsCapability -Online -Name OpenSSH.Client*
```
Or, alternatively, you could use e.g. [Putty](https://www.chiark.greenend.org.uk/~sgtatham/putty/).

Authenticate with the gateway:
```powershell
$gateway = "GARDENA-123456"
$password = "1234abcd"

$body = @{ password = $password } | ConvertTo-Json

add-type @"
    using System.Net;
    using System.Security.Cryptography.X509Certificates;
    public class TrustAllCertsPolicy : ICertificatePolicy {
        public bool CheckValidationResult(
            ServicePoint srvPoint, X509Certificate certificate,
            WebRequest request, int certificateProblem) { return true; }
    }
"@
[System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy

$response = Invoke-RestMethod -Method Post -Uri "https://$gateway/login" -Body $body -ContentType "application/json"
$session = $response.session
```

Register your public key:
```powershell
$key = Get-Content "$env:USERPROFILE\.ssh\id_ed25519.pub" -Raw
$body = @{ key = $key.Trim() } | ConvertTo-Json

Invoke-RestMethod -Method Post -Uri "https://$gateway/ssh_access_credentials" -Headers @{ "X-session" = $session } -Body $body -ContentType "application/json"
```

Enable SSH access:
```powershell
$body = @{ enable = $true } | ConvertTo-Json

Invoke-RestMethod -Method Put -Uri "https://$gateway/ssh_access_enable" -Headers @{ "X-session" = $session } -Body $body -ContentType "application/json"
```

Log into the gateway:
```txt
ssh root@$gateway
```

The instruction for enabling the WebSocket API can be found in [ha-gardena-smart-local-preview's README](https://github.com/cloudless-garden/ha-gardena-smart-local-preview/blob/main/README.md).

If you wait another 3 weeks, (hopefully) using SSH will no longer be necessary. You will be able to directly enable the WebSocket service with commands like above. We will update the [README](https://github.com/cloudless-garden/ha-gardena-smart-local-preview/blob/main/README.md) when the new image has been rolled out.

**Comment by andrexp** [comment]
Oh my goodness!!

It works. Finally. I installed the HA integration and all is automatically done. I am so lucky!!! You made my day and my weekend!

Thank you so much!

**Comment by easybe** [comment]
> Oh my goodness!!
> 
> It works. Finally. I installed the HA integration and all is automatically done. I am so lucky!!! You made my day and my weekend!
> 
> Thank you so much!

Great to hear! Let us know if things can be improved and of course, PRs welcome!

**Comment by andrexp** [comment]
Today is the first time my mower operates with the functional local connection to HA. I have an automation which is responsible to open and close a shutter of the mower garage as long as the status become "docked". It worked like charm except for one time (it resulted in an error "dock is blocked"). The reason is unknown. I will report issues as soon as I find any. There is maybe also a fault in my automation rule.

Cheers and have a nice weekend!

**Comment by lsiepel** [comment]
Was looking into some openHAB related requests and noticed this is moving forward. I lack the knowledge and details, but do i understand correctly that enabvling local access is now much easier? And the connection itself is now using websockets?

If so we could look into integrating this into openHAB. Related enhancement request: https://github.com/openhab/openhab-addons/issues/14904
